clover.rb

```diff
@@ -342,19 +348,41 @@ def assets
     password_confirm_label "Password Confirmation"
     before_create_account do
       Validation.validate_cloudflare_turnstile(param("cf-turnstile-response"))
-      account[:id] = Account.generate_uuid
-      account[:name] = param("name")
-      Validation.validate_account_name(account[:name])
+      scope.before_rodauth_create_account(account, param("name"))
     end
     after_create_account do
-      account = Account[account_id]
-      account.create_project_with_default_policy("Default")
-      ProjectInvitation.where(email: account.email).each do |inv|
-        account.associate_with_project(inv.project)
-        if (managed_policy = Authorization::ManagedPolicy.from_name(inv.policy))
-          managed_policy.apply(inv.project, [account], append: true)
+      scope.after_rodauth_create_account(account_id)
+    end
+
+    # :nocov:
+    if Config.omniauth_github_id
+      require "omniauth-github"
+      omniauth_provider :github, Config.omniauth_github_id, Config.omniauth_github_secret
+    end
+    if Config.omniauth_google_id
+      require "omniauth-google-oauth2"
+      omniauth_provider :google_oauth2, Config.omniauth_google_id, Config.omniauth_google_secret, name: :google
+    end
+    # :nocov:
+
+    before_omniauth_create_account do
+      scope.before_rodauth_create_account(account, omniauth_name)
+    end
+
+    after_omniauth_create_account do
+      scope.after_rodauth_create_account(account_id)
+    end
+
+    before_omniauth_callback_route do
+      account = Account[account_from_omniauth&.dig(:id)]
+      if authenticated?
+        if (account && account.id != scope.current_account.id) || !omniauth_email.casecmp?(scope.current_account.email)
+          flash["error"] = "Your account's email address is different from the email address associated with the #{omniauth_provider.capitalize} account."
+          redirect "/account/login-method"
         end
-        inv.destroy
+      elsif account && account.identities_dataset.where(provider: omniauth_provider.to_s).empty?
+        flash["error"] = "This account is already exists with another provider. Please login with the existing account and link the new provider from your account settings."
```

Comment:
I don't think the wording here is correct.  You get this message when there is an existing account for the email not tied to the social login, and you try to use social login for it.  Maybe:

```suggestion
        flash["error"] = "There is already an account with this email address, and it has not been linked to the #{omniauth_provider.capitalize} account. Please login to the existing account normally, and then link it to the #{omniauth_provider.capitalize} account from your account settings.  Then you can can login using the #{omniauth_provider.capitalize} account."
```

Reply:
Thanks, applied it.